### PR TITLE
Adds necessary constraints to database fields

### PR DESCRIPTION
**Pull request by PuneetPunamiya**
This adds necessary constraints to db fields and tests to validate them.
Also moves all migrations to their own files.

### CODE REVIEW

File: api/pkg/db/model/model.go [hunk 1]
```diff
@@ -21,19 +21,19 @@ type (
 
 	Catalog struct {
 		gorm.Model
-		Name       string
-		Type       string
-		URL        string
-		Owner      string
+		Name       string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:24]: should we set a size?

File: api/pkg/db/model/model.go [hunk 2]
```diff
@@ -21,19 +21,19 @@ type (
...
+		Type       string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:25]: why is the type unique?

File: api/pkg/db/model/model.go [hunk 3]
```diff
@@ -21,19 +21,19 @@ type (
...
+		URL        string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:26]: why is this unique? wouldn't this prevent 2 branches of the same repo to be added a s catalog?

Comment (sthaha) [reply api/pkg/db/model/model.go:26]: IMHO, we should only add constraints that are absolutely required by the domain we are trying to map.

File: api/pkg/db/model/model.go [hunk 4]
```diff
@@ -21,19 +21,19 @@ type (
...
+		Owner      string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:27]: I think catalog owner can also repeat, so `unique` isn't correct.

Comment (sthaha) [inline api/pkg/db/model/model.go:27]: Do we need this field? Can we remove this until a feature requires it? I am also thinking that the mapping may be one to many; i.e. a catalog is owned by multiple owners ..

File: api/pkg/db/model/model.go (hunk 1 above)
Comment (SM43) [reply api/pkg/db/model/model.go:24]: i am confused whether to set or size, by default it is 255

File: api/pkg/db/model/model.go [hunk 5]
```diff
@@ -21,19 +21,19 @@ type (
...
 		Resources  []Resource
-		Revision   string
+		Revision   string `gorm:"size:100;not null;unique"`
 	}
 
 	Resource struct {
 		gorm.Model
-		Name      string
-		Type      string
+		Name      string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:35]: name isn't unique but `catalog - name` combination is

File: api/pkg/db/model/model.go [hunk 6]
```diff
@@ -21,19 +21,19 @@ type (
...
+		Type      string `gorm:"size:100;not null;unique"`
```
Comment (sthaha) [inline api/pkg/db/model/model.go:36]: it isn't unique; see above

File: api/pkg/db/model/model.go (hunk 5 above)
Comment (SM43) [inline api/pkg/db/model/model.go:35]: resource name should not be unique, can repeat. we can have multiple catalogs

File: api/pkg/db/model/model.go (hunk 6 above)
Comment (SM43) [inline api/pkg/db/model/model.go:36]: Type also can't be unique

File: api/pkg/db/model/model.go [hunk 7]
```diff
@@ -43,11 +43,11 @@ type (
 
 	ResourceVersion struct {
 		gorm.Model
-		Version             string
+		Version             string `gorm:"size:100;not null;unique"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:46]: Version can't be unique eg. all the resources will have `v0.1`

File: api/pkg/db/model/model.go [hunk 8]
```diff
@@ -43,11 +43,11 @@ type (
...
 		Description         string
-		URL                 string
+		URL                 string `gorm:"size:100;not null;unique"`
 		DisplayName         string
-		MinPipelinesVersion string
+		MinPipelinesVersion string `gorm:"size:100;not null;unique"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:50]: MinPipelinesVersion will repeat so can't be unique

File: api/cmd/db/migration.go [hunk 9]
```diff
@@ -18,6 +18,54 @@ func Migrate(api *app.APIConfig) error {
 		api.DB(),
 		gormigrate.DefaultOptions,
 		[]*gormigrate.Migration{
+			{
+				ID: "202007141723",
+				Migrate: func(tx *gorm.DB) error {
+
```
Comment (SM43) [inline api/cmd/db/migration.go:24]: All queries can be written under one migration

File: api/cmd/db/migration.go [hunk 10]
```diff
@@ -122,14 +170,14 @@ func fkey(log *zap.SugaredLogger, db *gorm.DB, model interface{}, args ...string
...
 	var categories = map[string][]string{
-		"Others":         []string{},
-		"Build Tools":    []string{"build-tool"},
-		"CLI":            []string{"cli"},
-		"Cloud":          []string{"gcp", "aws", "azure", "cloud"},
-		"Deploy":         []string{"deploy"},
-		"Image Build":    []string{"image-build"},
-		"Notification":   []string{"notification"},
-		"Test Framework": []string{"test"},
+		"Others":         {},
```
Comment (SM43) [inline api/cmd/db/migration.go:173]: why this change?

File: api/pkg/db/model/model.go [hunk 11]
```diff
@@ -7,33 +7,33 @@ import (
...
 	Catalog struct {
 		gorm.Model
-		Name       string
-		Type       string
-		URL        string
-		Owner      string
+		Name       string `gorm:"not null"`
+		Type       string `gorm:"not null"`
+		URL        string `gorm:"not null"`
+		Owner      string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:27]: you can remove this field as discussed before.

Comment (SM43) [reply api/pkg/db/model/model.go:27]: Owner

File: api/pkg/db/model/model.go [hunk 12]
```diff
@@ -7,33 +7,33 @@ import (
...
 		Resources  []*Resource `gorm:"many2many:resource_tags;"`
 	}
 
 	Catalog struct {
 		gorm.Model
-		Name       string
-		Type       string
-		URL        string
-		Owner      string
+		Name       string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:24]: `catalog - name - type` combination will be unique if you can add a constraint for that, would be nice

File: api/pkg/db/migration/migration_test.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+func TestMigrateCheckNotNull(t *testing.T) {
+	tc := testutils.Setup(t)
+
+	err := tc.Error()
+
+	db := tc.DB()
+
+	err = db.Create(&model.Catalog{Name: "tekton", Type: "", URL: "", Revision: "master"}).Error
```
Comment (SM43) [inline api/pkg/db/migration/migration_test.go:18]: `assert` should be for each query

File: api/pkg/db/migration/migration_test.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"github.com/tektoncd/hub/api/pkg/testutils"
+)
+
+func TestMigrateCheckNotNull(t *testing.T) {
+	tc := testutils.Setup(t)
+
+	err := tc.Error()
```
Comment (SM43) [inline api/pkg/db/migration/migration_test.go:14]: why this is needed?

File: api/pkg/db/model/model.go [hunk 15]
```diff
@@ -21,33 +21,32 @@ import (
 type (
 	Category struct {
 		gorm.Model
-		Name string `gorm:"size:100;not null;unique"`
+		Name string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:24]: There can't be 2 categories with same name. So, unique can be applied here.

File: api/pkg/db/model/model.go [hunk 16]
```diff
@@ -21,33 +21,32 @@ import (
...
 		Tags []Tag
 	}
 
 	Tag struct {
 		gorm.Model
-		Name       string `gorm:"size:100;not null;unique"`
+		Name       string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:30]: Similarly here also there can't be 2 tags with same name. So, unique can be applied here.

**Comment by SM43** [comment]
@PuneetPunamiya There is no Migration written. Does this case not need one?

**Comment by PuneetPunamiya** [comment]
> @PuneetPunamiya There is no Migration written. Does this case not need one?

Done !

File: api/pkg/db/model/model.go [hunk 17]
```diff
@@ -21,33 +21,32 @@ import (
...
 		CategoryID uint
 		Resources  []*Resource `gorm:"many2many:resource_tags;"`
 	}
 
 	Catalog struct {
 		gorm.Model
-		Name       string
-		Type       string
-		URL        string
-		Owner      string
```
Comment (SM43) [inline api/pkg/db/model/model.go:41]: Removing this field will not delete the column, you will need to delete in the migration

File: api/cmd/db/migration.go [hunk 18]
```diff
@@ -32,6 +32,50 @@ func Migrate(api *app.APIConfig) error {
...
+					}
+					if err := tx.Exec(resource_version_query).Error; err != nil {
+						log.Error(err)
+						return err
+					}
+					return nil
+				},
+			},
+			{
+				ID: "202005080741",
```
Comment (SM43) [inline api/cmd/db/migration.go:70]: Why a separate migration for this?

File: api/cmd/db/migration.go [hunk 19]
```diff
@@ -32,6 +32,44 @@ func Migrate(api *app.APIConfig) error {
 		api.DB(),
 		gormigrate.DefaultOptions,
 		[]*gormigrate.Migration{
+			{
+				ID: "202007141723",
```
Comment (SM43) [inline api/cmd/db/migration.go:36]: Could you please add a comment above `ID` what the migration does?

File: api/pkg/db/model/model.go [hunk 20]
```diff
@@ -21,33 +21,32 @@ import (
...
+		Name       string `gorm:"primary_key"`
+		Type       string `gorm:"primary_key"`
+		URL        string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:40]: Does this not requires `default:null`?

File: api/pkg/db/validation/validation_test.go [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"github.com/tektoncd/hub/api/pkg/testutils"
+)
+
+func TestCheckNotNull(t *testing.T) {
+	tc := testutils.Setup(t)
+
+	err := tc.Error()
```
Comment (SM43) [inline api/pkg/db/validation/validation_test.go:30]: Is this required here?

File: api/pkg/db/validation/validation_test.go [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+	db := tc.DB()
+
+	err = db.Create(&model.Catalog{Name: "tekton", Type: "", URL: "", Revision: "master"}).Error
```
Comment (SM43) [inline api/pkg/db/validation/validation_test.go:34]: I think this will return an error only because `Type` is null and we won't know whether `URL` is `NOT NULL` or not.

File: api/pkg/db/model/model.go [hunk 23]
```diff
@@ -21,33 +21,32 @@ import (
...
 		ContextDir string
 		Resources  []Resource
-		Revision   string
+		Revision   string `gorm:"not null"`
```
Comment (SM43) [inline api/pkg/db/model/model.go:43]: Using only `NOT NULL` won't be helpful I guess, we also need `default:null`? WDYT?

File: api/pkg/db/initializer/initiailizer.go [hunk 24]
```diff
@@ -81,7 +81,6 @@ func (i *Initializer) addCatalogs() error {
 			Name:       c.Name,
 			Type:       c.Type,
 			URL:        c.URL,
-			Owner:      c.Owner,
```
Comment (SM43) [inline api/pkg/db/initializer/initiailizer.go:84]: Could you also remove `owner` from config.yaml and `initializer`?

File: api/pkg/db/migration/202006071000_add_foreign_Key_constraints.go [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"github.com/jinzhu/gorm"
+	"go.uber.org/zap"
+	"gopkg.in/gormigrate.v1"
+
+	"github.com/tektoncd/hub/api/pkg/db/model"
+)
+
+func addForeignKeyConstraints(log *zap.SugaredLogger) *gormigrate.Migration {
```
Comment (SM43) [inline api/pkg/db/migration/202006071000_add_foreign_Key_constraints.go:25]: this migration adds the resource table and then adds the `fk` so I think the name should be something related to creating tables.

File: api/pkg/db/migration/202008071700_add_migration_on_catalog.go [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"github.com/jinzhu/gorm"
+	"go.uber.org/zap"
+	"gopkg.in/gormigrate.v1"
+
+	"github.com/tektoncd/hub/api/pkg/db/model"
+)
+
+func addMigrationOnCatalog(log *zap.SugaredLogger) *gormigrate.Migration {
```
Comment (SM43) [inline api/pkg/db/migration/202008071700_add_migration_on_catalog.go:25]: I think Migration don't need to be in the name, could be what the migration does?
